v1/ocagent.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ocagent
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+
+	"go.opencensus.io/trace"
+
+	agentcommonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	agenttracepb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/v1"
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+)
+
+var startupMu sync.Mutex
+var startTime time.Time
+
+func init() {
+	startupMu.Lock()
+	startTime = time.Now()
+	startupMu.Unlock()
+}
+
+var _ trace.Exporter = (*Exporter)(nil)
+
+type Exporter struct {
+	mu              sync.RWMutex
+	started         bool
+	agentPort       int
+	agentHost       string
+	canDialInsecure bool
+	traceSvcClient  agenttracepb.TraceServiceClient
+	traceExporter   agenttracepb.TraceService_ExportClient
+	nodeInfo        *agentcommonpb.Node
+}
+
+func NewExporter(opts ...ExporterOption) (*Exporter, error) {
+	e := new(Exporter)
+	for _, opt := range opts {
+		opt.withExporter(e)
+	}
+	if e.agentPort <= 0 {
+		e.agentPort = DefaultAgentPort
+	}
+	if e.agentHost == "" {
+		e.agentHost = DefaultAgentHost
+	}
+	e.nodeInfo = createNodeInfo("foo")
+	return e, nil
+}
+
+func (ae *Exporter) Start() error {
+	ae.mu.Lock()
+	defer ae.mu.Unlock()
+
+	if ae.started {
+		return nil
+	}
+
+	// Now start it
+	ae.started = true
+	addr := fmt.Sprintf("%s:%d", ae.agentHost, ae.agentPort)
+	var dialOpts []grpc.DialOption
+	if ae.canDialInsecure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+	cc, err := grpc.Dial(addr, dialOpts...)
+	if err != nil {
+		return err
+	}
+
+	traceSvcClient := agenttracepb.NewTraceServiceClient(cc)
+	traceExporter, err := traceSvcClient.Export(context.Background())
+	if err != nil {
+		return fmt.Errorf("Exporter.Start:: TraceServiceClient: %v", err)
+	}
+	// First things first, send over the nodeID information
+	configStream, err := traceSvcClient.Config(context.Background())
+	if err != nil {
+		return fmt.Errorf("Exporter.Start:: ConfigStream: %v", err)
+	}
+
+	// Initiate the config service
```

Comment:
We need to do this again in case of an error on the config/export RPCs.

Reply:
Done in the last commit, I've added exponential backoff too, PTAL.